subsys/net/lib/nrf_cloud/coap/src/coap_codec.c

```diff
@@ -0,0 +1,631 @@
+	__ASSERT_NO_MSG(buf != NULL);
+	__ASSERT_NO_MSG(len != NULL);
+
+	if (fmt != COAP_CONTENT_FORMAT_APP_CBOR) {
+		LOG_ERR("Invalid format for ground fix: %d", fmt);
+		return -ENOTSUP;
+	}
+
+	int err = 0;
+	struct ground_fix_req input;
+	size_t out_len;
+
+	memset(&input, 0, sizeof(struct ground_fix_req));
+	input._ground_fix_req_lte_present = (cell_info != NULL);
+	if (cell_info) {
+		copy_cell_info(&input._ground_fix_req_lte._ground_fix_req_lte, cell_info);
+	}
+	input._ground_fix_req_wifi_present = (wifi_info != NULL);
+	if (wifi_info) {
+		copy_wifi_info(&input._ground_fix_req_wifi._ground_fix_req_wifi, wifi_info);
+	}
+	err = cbor_encode_ground_fix_req(buf, *len, &input, &out_len);
+	if (err) {
+		LOG_ERR("Error %d encoding groundfix", err);
+		err = -EINVAL;
+		*len = 0;
+	} else {
+		*len = out_len;
+	}
+	return err;
+}
+
+int coap_codec_ground_fix_resp_decode(struct nrf_cloud_location_result *result,
+				      const uint8_t *buf, size_t len, enum coap_content_format fmt)
+{
+	__ASSERT_NO_MSG(result != NULL);
+	__ASSERT_NO_MSG(buf != NULL);
+	int err;
+
+	if (fmt == COAP_CONTENT_FORMAT_APP_JSON) {
+		enum nrf_cloud_error nrf_err;
+
+		/* Check for a potential ground fix JSON error message from nRF Cloud */
+		err = nrf_cloud_error_msg_decode(buf, NRF_CLOUD_JSON_APPID_VAL_LOCATION,
+						 NRF_CLOUD_JSON_MSG_TYPE_VAL_DATA, &nrf_err);
+		if (!err) {
+			LOG_ERR("nRF Cloud returned A-GPS error: %d", nrf_err);
+			return -EFAULT;
+		}
+		LOG_ERR("Invalid A-GPS response format: %d", err);
+		return -EPROTO;
+	}
+	if (fmt != COAP_CONTENT_FORMAT_APP_CBOR) {
+		LOG_ERR("Invalid format for ground fix: %d", fmt);
+		return -ENOTSUP;
+	}
+
+	struct ground_fix_resp res;
+	size_t out_len;
+
+	err = cbor_decode_ground_fix_resp(buf, len, &res, &out_len);
+	if (out_len != len) {
+		LOG_WRN("Different response length: expected:%zd, decoded:%zd",
+			len, out_len);
+	}
+	if (err) {
+		return err;
+	}
+
+	const char *with = res._ground_fix_resp_fulfilledWith.value;
+	int rlen = res._ground_fix_resp_fulfilledWith.len;
+
+	if (strncmp(with, NRF_CLOUD_LOCATION_TYPE_VAL_MCELL, rlen) == 0) {
+		result->type = LOCATION_TYPE_MULTI_CELL;
+	} else if (strncmp(with, NRF_CLOUD_LOCATION_TYPE_VAL_SCELL, rlen) == 0) {
+		result->type = LOCATION_TYPE_SINGLE_CELL;
+	} else if (strncmp(with, NRF_CLOUD_LOCATION_TYPE_VAL_WIFI, rlen) == 0) {
+		result->type = LOCATION_TYPE_WIFI;
+	} else {
+		result->type = LOCATION_TYPE__INVALID;
+		LOG_WRN("Unhandled location type: %*s", rlen, with);
+	}
+
+	result->lat = res._ground_fix_resp_lat;
+	result->lon = res._ground_fix_resp_lon;
+
+	if (res._ground_fix_resp_uncertainty_choice == _ground_fix_resp_uncertainty_int) {
+		result->unc = res._ground_fix_resp_uncertainty_int;
+	} else {
+		result->unc = (uint32_t)lround(res._ground_fix_resp_uncertainty_float);
+	}
+	return err;
+}
+
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+static int format_agps_custom_types_array(struct nrf_modem_gnss_agps_data_frame const *const req,
```

Comment:
is this the same as `nrf_cloud_agps_type_array_get()`?

Reply:
Now it is.  I'm not sure it existed when I wrote this originally.